feat: fix schema issues for `native reader - read STRUCT of ARRAY fields`

Followup ticket to fix schema issues for `native reader - read STRUCT of ARRAY fields`

_Originally posted by @comphead in https://github.com/apache/datafusion-comet/pull/1550#discussion_r2002258552_
            

Related to #1454

fyi, currently some tests in Spark SQL (e.g. `"add a nested column at the end of the leaf struct column"`) fail because of structs like 
```
 struct
   +- array
           +- struct
```
The tests fail with 
```
Caused by: org.apache.comet.CometNativeException: no entry found for key
[info]         at comet::errors::init::{{closure}}(__internal__:0)
[info]         at std::panicking::rust_panic_with_hook(__internal__:0)
[info]         at std::panicking::begin_panic_handler::{{closure}}(__internal__:0)
[info]         at std::sys::backtrace::__rust_end_short_backtrace(__internal__:0)
[info]         at rust_begin_unwind(__internal__:0)
[info]         at core::panicking::panic_fmt(__internal__:0)
[info]         at core::option::expect_failed(__internal__:0)
[info]         at comet::parquet::parquet_support::cast_struct_to_struct(__internal__:0)
[info]         at comet::parquet::parquet_support::cast_array(__internal__:0)
[info]         at comet::parquet::parquet_support::cast_struct_to_struct(__internal__:0)
[info]         at comet::parquet::parquet_support::cast_array(__internal__:0)
[info]         at comet::parquet::parquet_support::spark_parquet_convert(__internal__:0)
[info]         at <core::iter::adapters::GenericShunt<I,R> as core::iter::traits::iterator::Iterator>::next(__internal__:0)
[info]         at <comet::parquet::schema_adapter::SchemaMapping as datafusion::datasource::schema_adapter::SchemaMapper>::map_batch(__internal__:0)
[info]         at <futures_util::stream::stream::map::Map<St,F> as futures_core::stream::Stream>::poll_next(__internal__:0)
[info]         at <datafusion_datasource::file_stream::FileStream as futures_core::stream::Stream>::poll_next(__internal__:0)
[info]         at Java_org_apache_comet_parquet_Native_readNextRecordBatch(__internal__:0)
[info]         at <unknown>(__internal__:0)
[info] 	at org.apache.comet.parquet.Native.readNextRecordBatch(Native Method)
```

--- a/native/core/src/execution/shuffle/row.rs
+++ b/native/core/src/execution/shuffle/row.rs
@@ -3198,58 +3198,61 @@ fn make_builders(
             let value_builder =
                 make_builders(field.data_type(), NESTED_TYPE_BUILDER_CAPACITY, 1.0)?;
 
+            // Needed to overwrite default ListBuilder creation having the incoming field schema to be driving
+            let value_field = Arc::clone(&field);
+
             match field.data_type() {
                 DataType::Boolean => {
                     let builder = downcast_builder!(BooleanBuilder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Int8 => {
                     let builder = downcast_builder!(Int8Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Int16 => {
                     let builder = downcast_builder!(Int16Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Int32 => {
                     let builder = downcast_builder!(Int32Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Int64 => {
                     let builder = downcast_builder!(Int64Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Float32 => {
                     let builder = downcast_builder!(Float32Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Float64 => {
                     let builder = downcast_builder!(Float64Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Decimal128(_, _) => {
                     let builder = downcast_builder!(Decimal128Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Timestamp(TimeUnit::Microsecond, _) => {
                     let builder = downcast_builder!(TimestampMicrosecondBuilder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Date32 => {
                     let builder = downcast_builder!(Date32Builder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Binary => {
                     let builder = downcast_builder!(BinaryBuilder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Utf8 => {
                     let builder = downcast_builder!(StringBuilder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 DataType::Struct(_) => {
                     let builder = downcast_builder!(StructBuilder, value_builder);
-                    Box::new(ListBuilder::new(*builder))
+                    Box::new(ListBuilder::new(*builder).with_field(value_field))
                 }
                 // TODO: nested list is not supported. Due to the design of `ListBuilder`, it has
                 // a `T: ArrayBuilder` as type parameter. It makes hard to construct an arbitrarily
@@ -3274,6 +3277,7 @@ fn make_builders(
                 .map(|field| make_builders(field.data_type(), row_num, 1.0))
                 .collect::<Result<Vec<_>, _>>()?;
 
+            dbg!(fields);
             Box::new(StructBuilder::new(fields.clone(), field_builders))
         }
         _ => return Err(CometError::Internal(format!("Unsupported type: {:?}", dt))),

--- a/native/core/src/execution/shuffle/row.rs
+++ b/native/core/src/execution/shuffle/row.rs
@@ -3199,7 +3199,7 @@ fn make_builders(
                 make_builders(field.data_type(), NESTED_TYPE_BUILDER_CAPACITY, 1.0)?;
 
             // Needed to overwrite default ListBuilder creation having the incoming field schema to be driving
-            let value_field = Arc::clone(&field);
+            let value_field = Arc::clone(field);
 
             match field.data_type() {
                 DataType::Boolean => {

--- a/native/core/src/execution/shuffle/row.rs
+++ b/native/core/src/execution/shuffle/row.rs
@@ -3277,7 +3277,6 @@ fn make_builders(
                 .map(|field| make_builders(field.data_type(), row_num, 1.0))
                 .collect::<Result<Vec<_>, _>>()?;
 
-            dbg!(fields);
             Box::new(StructBuilder::new(fields.clone(), field_builders))
         }
         _ => return Err(CometError::Internal(format!("Unsupported type: {:?}", dt))),

--- a/spark/src/test/scala/org/apache/comet/exec/CometNativeReaderSuite.scala
+++ b/spark/src/test/scala/org/apache/comet/exec/CometNativeReaderSuite.scala
@@ -62,25 +62,7 @@ class CometNativeReaderSuite extends CometTestBase with AdaptiveSparkPlanHelper 
       "select arr from tbl")
   }
 
-  /*
-  native reader - read STRUCT of ARRAY fields - native_datafusion *** FAILED *** (191 milliseconds)
-  org.apache.spark.SparkException: Job aborted due to stage failure: Task 1 in stage 29.0 failed 1 times,
-  most recent failure: Lost task 1.0 in stage 29.0 (TID 35) (192.168.4.142 executor driver):
-  org.apache.comet.CometNativeException: called `Result::unwrap()` on an `Err` value:
-  InvalidArgumentError("Incorrect datatype for StructArray field \"col\", expected List(Field { name: \"item\",
-  data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} })
-  got List(Field { name: \"item\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })")
-
-
-  native reader - read STRUCT of ARRAY fields - native_iceberg_compat *** FAILED *** (82 milliseconds)
-  org.apache.spark.SparkException: Job aborted due to stage failure:
-  Task 1 in stage 31.0 failed 1 times, most recent failure: Lost task 1.0 in stage 31.0 (TID 39) (192.168.4.142 executor driver):
-  org.apache.comet.CometNativeException: called `Result::unwrap()` on an `Err` value: InvalidArgumentError("Incorrect datatype for StructArray field \"col\",
-  expected List(Field { name: \"item\", data_type: Int32, nullable: false, dict_id: 0, dict_is_ordered: false, metadata: {} })
-  got List(Field { name: \"item\", data_type: Int32, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })")
-
-   */
-  ignore("native reader - read STRUCT of ARRAY fields") {
+  test("native reader - read STRUCT of ARRAY fields") {
     testSingleLineQuery(
       """
         |select named_struct('col', arr) c0 from

